temporal-sdk/src/main/java/io/temporal/worker/tuning/SlotPermit.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.worker.tuning;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This class is handed out by implementations of {@link SlotSupplier}. Permits are held until the
+ * tasks they are associated with (if any) are finished processing, or if the reservation is no
+ * longer needed. Your supplier implementation may store additional data in the permit, if desired.
+ */
+public class SlotPermit {
+  public final long id;
+  public final Object userData;
+
+  private static final AtomicLong nextId = new AtomicLong(0);
+
+  public SlotPermit() {
+    id = nextId.getAndIncrement();
+    this.userData = null;
+  }
+
+  public SlotPermit(Object userData) {
```

Comment:
I suspect slot providers will often be implemented by chaining  or wrapping other slot providers, where each slot provider may want to attach data to the permit. Maybe an API like a Go `context`?

Reply:
Are you thinking something like `permit.withData(obj)`?

Reply:
This is where allowing the Java author to return the type of their choice (even if they were forced to extend `SlotPermit` though not as much of a fan there) gives more freedom than attaching an object to an existing object.

Reply:
I would not want users to be able to return an object of their choice. I would like to maintain control over the object and provide an interface to attach and get metadata.

Reply:
:+1: Ok, no prob. Still not a big fan of the ID in here, but no prob there too. Mentioned in other comment, but we should probably make this class final, consider removing the equals/hashCode override (i.e. default to identity-based), and document on the context interfaces that the permits are the _exact same instances_ as the ones returned.

Reply:
Fixed to do that

Reply:
To clarify I was taking about having a map of userdata like a Go `context` does so different providers wrapping each other can add their own metadata without impacting each other.

Reply:
Something like a typemap of <T -> Data> would be useful. But, perhaps we add that later if requested